compiler/rustc_parse/src/parser/item.rs

```diff
@@ -59,10 +59,18 @@ impl<'a> Parser<'a> {
 
         let post_attr_lo = self.token.span;
         let mut items = ThinVec::new();
-        while let Some(item) = self.parse_item(ForceCollect::No)? {
-            self.maybe_consume_incorrect_semicolon(Some(&item));
+
+        // There shouldn't be any semicolons before or afer items.
+        // `parse_item` consumes the appropriate semicolons so any leftover is an error.
+        let mut last = None;
+        while let Some(item) = {
+            self.maybe_consume_incorrect_semicolon(last);
+            self.parse_item(ForceCollect::No)?
+        } {
             items.push(item);
+            last = items.last().map(|x| &**x);
         }
```

Comment:
Will this work when you have multiple consecutive incorrect semicolons e.g.:
```rust
;
;
fn main() { }
```

I think you should call `maybe-consume_incorrect_semicolon` in a loop and consume as many of them as possible before calling `parse_item`

Reply:
Good catch